packages/edit-post/src/components/manage-blocks-modal/manager.js

```diff
@@ -0,0 +1,133 @@
+/**
+ * External dependencies
+ */
+import { filter, without } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { withSelect } from '@wordpress/data';
+import { compose } from '@wordpress/compose';
+import { Component } from '@wordpress/element';
+import { TextControl } from '@wordpress/components';
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import BlockManagerCategory from './category';
+
+class BlockManager extends Component {
+	constructor() {
+		super( ...arguments );
+
+		this.togglePanel = this.togglePanel.bind( this );
+		this.setSearch = this.setSearch.bind( this );
+
+		this.state = {
+			search: '',
+			openPanels: [],
+		};
+	}
+
+	/**
+	 * Given a category slug, toggles its results panel as opened or closed.
+	 *
+	 * @param {string} category Category slug
+	 */
+	togglePanel( category ) {
+		let { openPanels } = this.state;
+
+		openPanels = openPanels.includes( category ) ?
+			without( openPanels.category ) :
+			openPanels.concat( category );
+
+		this.setState( { openPanels } );
+	}
+
+	/**
+	 * Sets the search term by which to filter results.
+	 *
+	 * @param {string} search Search term.
+	 */
+	setSearch( search ) {
+		this.setState( { search } );
+	}
+
+	render() {
+		const {
+			categories,
+			blockTypes,
+			hiddenBlockTypes,
+			hasBlockSupport,
+			isMatchingSearchTerm,
+		} = this.props;
+		const { search, openPanels } = this.state;
+
+		const blockItems = blockTypes
+			.filter( ( blockType ) => (
+				hasBlockSupport( blockType, 'inserter', true ) &&
+				( ! search || isMatchingSearchTerm( blockType, search ) )
+			) )
+			.map( ( blockType ) => ( {
+				id: blockType.name,
+				icon: blockType.icon,
+				title: blockType.title,
+				category: blockType.category,
+			} ) );
+
+		return (
+			<div className="edit-post-manage-blocks-modal__content">
+				<TextControl
+					type="search"
+					label={ __( 'Search for a block' ) }
+					value={ search }
+					onChange={ this.setSearch }
+					className="edit-post-manage-blocks-modal__search"
+				/>
+				<div
+					tabIndex="0"
+					role="region"
+					aria-label={ __( 'Available block types' ) }
+					className="edit-post-manage-blocks-modal__results"
+				>
+					{ categories.map( ( category ) => (
```

Comment:
We still don't handle the case where there are no blocks found after filters get applied:

![Screen Shot 2019-03-12 at 13 05 00](https://user-images.githubusercontent.com/699132/54198713-8316c900-44c7-11e9-9c91-3da8a8d41c0b.png)

Reply:
> We still don't handle the case where there are no blocks found after filters get applied:

Added a "No Results" message in 6171177.